### PR TITLE
User register complete & user login validation complete, but without session 

### PR DESCRIPTION
**Pull request by s3c0nDD**

### CODE REVIEW

File: controllers/articles/controller.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ func (c *ArticleController) AddArticle(w http.ResponseWriter, r *http.Request) {
 
 // SaveArticle is handler to save article from form obtained data
 func (c *ArticleController) SaveArticle(w http.ResponseWriter, r *http.Request) {
+	log.Println("responding to", r.Method, r.URL)
```
Comment (SebastianPozoga) [inline controllers/articles/controller.go]: don't add logs lke that to repo

File: controllers/articles/controller.go [hunk 2]
```diff
@@ -85,11 +85,10 @@ func (c *ArticleController) ListArticle(w http.ResponseWriter, r *http.Request)
 func (c *ArticleController) ViewArticle(w http.ResponseWriter, r *http.Request) {
 	log.Println("responding to", r.Method, r.URL)
 	vars := mux.Vars(r)
-
 	articleID, _ := strconv.Atoi(vars["id"])
+	// Find article in database by given id
```
Comment (SebastianPozoga) [inline controllers/articles/controller.go]: redundant comment

File: controllers/articles/controller.go [hunk 3]
```diff
@@ -85,11 +85,10 @@ func (c *ArticleController) ListArticle(w http.ResponseWriter, r *http.Request)
...
 	article := c.articleDAO.FindByID(articleID)
-
-	if article == nil { // if fe. user gives id of non existent article
+	if article == nil { // if article of given ID doesn't exist
 		http.Error(w, http.StatusText(404), 403)
 		// TODO maybe handle above some better way?
```
Comment (SebastianPozoga) [inline controllers/articles/controller.go]: repo is not a chat. Don't ask here

File: controllers/users/controller.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		return
+	}
+	// encrypt password with bcrypt
+	passHashed, err := c.crypt.Hash(registerForm.Password)
+	if err != nil {
+		log.Fatal("error crypting pass: ", err)
+		return
+	}
+	user := usermodel.UserDTO{Email: registerForm.Email, PassHash: passHashed}
+	// ...and save to database
```
Comment (SebastianPozoga) [inline controllers/users/controller.go]: redundant comment

File: controllers/users/controller.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	var userToAdd []models.UserDTO
+	userToAdd = append(userToAdd, models.UserDTO(&user))
+	c.userDAO.PersistAll(userToAdd)
+	// redirect
```
Comment (SebastianPozoga) [inline controllers/users/controller.go]: redundant comment

File: forms/login.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/goatcms/goatcms/services"
+)
+
+// LoginForm is structure with login form values
+type LoginForm struct {
+	Email    string `schema:"email"`
+	Password string `schema:"password"`
+}
+
+var (
```
Comment (SebastianPozoga) [inline forms/login.go]: Should it be a const?

File: forms/register.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/goatcms/goatcms/models"
+)
+
+// RegisterForm is structure with register form values
+type RegisterForm struct {
+	Email    string `schema:"email"`
+	Password string `schema:"password"`
+}
+
+var (
```
Comment (SebastianPozoga) [inline forms/register.go]: const?

File: models/user/DAO.go [hunk 8]
```diff
@@ -42,6 +42,29 @@ func (dao *UserDAO) FindAll() []models.UserDTO {
 	return result
 }
 
+// FindByEmail obtain user of given email from database
+func (dao *UserDAO) FindByEmail(email string) models.UserDTO {
+	query := `
+		SELECT id, email, pass_hash FROM users
```
Comment (SebastianPozoga) [inline models/user/DAO.go]: email should be unique. You can add 
... LIMIT 1;
to your sql

File: models/user/DAO.go [hunk 9]
```diff
@@ -42,6 +42,29 @@ func (dao *UserDAO) FindAll() []models.UserDTO {
...
+		WHERE email = "` + email + `"`
+	rows, err := dao.db.Adapter().Query(query)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer rows.Close()
+
+	var result models.UserDTO
+	for rows.Next() {
```
Comment (SebastianPozoga) [inline models/user/DAO.go]: the for is illogical

File: services/crypt/main_test.go [hunk 10]
```diff
@@ -22,18 +22,51 @@ func TestIfHashAndPassAreEqual(t *testing.T) {
...
-func TestIfHashAndPassNotMatchingAreNotEqual(t *testing.T) {
-	const testPassword = "foobar"
+func TestIfHashAndWrongPassDontMatch(t *testing.T) {
 	cryptService := &Crypt{}
-	hashed, err := cryptService.Hash(testPassword)
+	hashed, err := cryptService.Hash("goodpassword")
 	if err != nil {
 		t.Error(err)
-		return
+		// return
```
Comment (SebastianPozoga) [inline services/crypt/main_test.go]: don't comment the return. It is to stop running crashed code

File: controllers/articles/controller.go (hunk 1 above)
Comment (s3c0nDD) [inline controllers/articles/controller.go]: so delete all logs like this?

File: controllers/users/controller.go (hunk 4 above)
Comment (s3c0nDD) [inline controllers/users/controller.go]: which one, there are plenty of them from lines 11 to 99 :cry:

File: forms/login.go (hunk 6 above)
Comment (s3c0nDD) [inline forms/login.go]: can be :)

File: forms/register.go (hunk 7 above)
Comment (s3c0nDD) [inline forms/register.go]: yup

File: models/user/DAO.go (hunk 9 above)
Comment (s3c0nDD) [inline models/user/DAO.go]: image

if i remove for like in image above, i get error like in image. so i'll leave it like it is until the better solution is made

File: services/crypt/main_test.go (hunk 10 above)
Comment (s3c0nDD) [inline services/crypt/main_test.go]: should i add 'return' to every block of:
if err != nil {
  // some code...
  return
}
in a testing funcs ?

File: models/user/DAO.go (hunk 9 above)
Comment (s3c0nDD) [inline models/user/DAO.go]: ok i'll change this for QueryRow

File: models/user/DAO.go (hunk 8 above)
Comment (s3c0nDD) [inline models/user/DAO.go]: it's correct now (look line 49), but on this line (48) diff show no change, so this comment didnt dissapear

File: models/user/DAO.go [hunk 11]
```diff
@@ -46,20 +47,17 @@ func (dao *UserDAO) FindAll() []models.UserDTO {
...
 	query := `
 		SELECT id, email, pass_hash FROM users
-		WHERE email = "` + email + `"`
-	rows, err := dao.db.Adapter().Query(query)
-	if err != nil {
-		log.Fatal(err)
-	}
-	defer rows.Close()
-
+		WHERE email = "` + email + `" LIMIT 1`
```
Comment (SebastianPozoga) [inline models/user/DAO.go]: http://stackoverflow.com/questions/26345318/how-can-i-prevent-sql-injection-attacks-in-go-while-using-database-sql
// this is safe
db.Query("SELECT name FROM users WHERE age=?", req.FormValue("age"))
// this allows sql injection.
db.Query("SELECT name FROM users WHERE age=" + req.FormValue("age"))

**Pull request merged**